Traffic Monitor Integration Test Framework

rob05c opened the pull request:
Adds a Traffic Monitor Integration Test Framework.

WIP, but making a draft for input. It needs some cleanup, but it works, and has a few tests. Instructions are in the README, let me know if you have trouble getting it to work.

Also adds a Go client for the Monitor (because it needed it).

Comments welcome.

<!--
************ STOP!! ************
If this Pull Request is intended to fix a security vulnerability, DO NOT submit it! Instead, contact
the Apache Software Foundation Security Team at security@trafficcontrol.apache.org and follow the
guidelines at https://www.apache.org/security/ regarding vulnerability disclosure.
-->
## What does this PR (Pull Request) do?
<!-- Explain the changes you made here. If this fixes an Issue, identify it by
replacing the text in the checkbox item with the Issue number e.g.

- [x] This PR fixes #9001 OR is not related to any Issue

^ This will automatically close Issue number 9001 when the Pull Request is
merged (The '#' is important).

Be sure you check the box properly, see the "The following criteria are ALL
met by this PR" section for details.
-->

- [ ] This PR fixes #REPLACE_ME OR is not related to any Issue <!-- You can check for an issue here: https://github.com/apache/trafficcontrol/issues -->


## Which Traffic Control components are affected by this PR?
<!-- Please delete all components from this list that are NOT affected by this
Pull Request. Also, feel free to add the name of a tool or script that is
affected but not on the list.

Additionally, if this Pull Request does NOT affect documentation, please
explain why documentation is not required. -->

- CDN in a Box
- Documentation
- Grove
- Traffic Control Client <!-- Please specify which; e.g. 'Python', 'Go', 'Java' -->
- Traffic Monitor
- Traffic Ops
- Traffic Ops ORT
- Traffic Portal
- Traffic Router
- Traffic Stats
- Traffic Vault

## What is the best way to verify this PR?
<!-- Please include here ALL the steps necessary to test your Pull Request. If
it includes tests (and most should), outline here the steps needed to run the
tests. If not, lay out the manual testing procedure and please explain why
tests are unnecessary for this Pull Request. -->

## If this is a bug fix, what versions of Traffic Ops are affected?
<!-- If this PR fixes a bug, please list here all of the affected versions - to
the best of your knowledge. It's also pretty helpful to include a commit hash
of where 'master' is at the time this PR is opened (if it affects master),
because what 'master' means will change over time. For example, if this PR
fixes a bug that's present in master (at commit hash '2697ebac'), in v3.0.0,
and in the current 3.0.1 Release candidate (e.g. RC1), then this list would
look like:

- master (2697ebac)
- 3.0.0
- 3.0.1 (RC1)

If you don't know what other versions might have this bug, AND don't know how
to find the commit hash of 'master', then feel free to leave this section
blank (or, preferably, delete it entirely).
 -->


## The following criteria are ALL met by this PR
<!-- Check the boxes to signify that the associated statement is true. To
"check a box", replace the space inside of the square brackets with an 'x'.
e.g.

- [ x] <- Wrong
- [x ] <- Wrong
- [] <- Wrong
- [*] <- Wrong
- [x] <- Correct!

-->

- [ ] This PR includes tests OR I have explained why tests are unnecessary
- [ ] This PR includes documentation OR I have explained why documentation is unnecessary
- [ ] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [ ] This PR includes any and all required license headers
- [ ] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [ ] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3918/
Test FAILed.

rob05c commented:
Rebased, removed draft/WIP. I haven't had time to do more with this, but it basically worked. I don't think there's a reason not to merge it, if anyone wants to use and extend it for Traffic Monitor Integration Testing.

zrhoffman commented on traffic_monitor/tests/integration/config/config.go:
```diff
@@ -0,0 +1,157 @@
+	if _, err := os.Stat(confPath); !os.IsNotExist(err) {
+		confBytes, err := ioutil.ReadFile(confPath)
+		if err != nil {
+			return Config{}, fmt.Errorf("Reading CDN conf '%s': %v", confPath, err)
```
*Reading* should not be capitalized

zrhoffman commented on traffic_monitor/tests/integration/config/config.go:
```diff
@@ -0,0 +1,157 @@
+	}
+	err := envconfig.Process("traffic-ops-client-tests", &cfg)
+	if err != nil {
+		fmt.Errorf("cannot parse config: %v\n", err)
```
Is this error meant to be printed?

zrhoffman commented on traffic_monitor/tests/integration/traffic_monitor_test.go:
```diff
@@ -0,0 +1,125 @@
+			   TM Session Timeout:   %ds
+`, *configFileName, Config.TrafficMonitor.URL, Config.Default.Session.TimeoutInSecs)
+
+	// //Load the test data
+	// LoadFixtures(*tcFixturesFileName)
+
+	// var db *sql.DB
+	// db, err = OpenConnection()
+	// if err != nil {
+	// 	fmt.Printf("\nError opening connection to %s - %s, %v\n", Config.TrafficOps.URL, Config.TrafficOpsDB.User, err)
+	// 	os.Exit(1)
+	// }
+	// defer db.Close()
+
+	// err = Teardown(db)
+	// if err != nil {
+	// 	fmt.Printf("\nError tearingdown data %s - %s, %v\n", Config.TrafficOps.URL, Config.TrafficOpsDB.User, err)
+	// 	os.Exit(1)
+	// }
+
+	// err = SetupTestData(db)
+	// if err != nil {
+	// 	fmt.Printf("\nError setting up data %s - %s, %v\n", Config.TrafficOps.URL, Config.TrafficOpsDB.User, err)
+	// 	os.Exit(1)
+	// }
```
This is all commented out because it's still *todo*, right? Could we get a `TODO` comment about this code block?

zrhoffman commented on traffic_monitor/tools/testcaches/fakesrvr/cmd.go:
```diff
@@ -0,0 +1,220 @@
+
+		srvr := (*fakesrvrdata.FakeServerData)(fakeSrvrDataThs.Get())
+		srvr.System.Speed = int(newSpeed)
+		fakeSrvrDataThs.Set(fakesrvrdata.ThsT(srvr))
```
This can be just

```go
fakeSrvrDataThs.Set(srvr)
```

zrhoffman commented on traffic_monitor/tools/testcaches/fakesrvr/cmd.go:
```diff
@@ -0,0 +1,220 @@
+
+		srvr := (*fakesrvrdata.FakeServerData)(fakeSrvrDataThs.Get())
+		srvr.System.ProcLoadAvg.CPU1m = newLoadAvg1M
+		fakeSrvrDataThs.Set(fakesrvrdata.ThsT(srvr))
```
This too can be

```go
fakeSrvrDataThs.Set(srvr)
```

zrhoffman commented on traffic_monitor/tools/testcaches/fakesrvr/cmd.go:
```diff
@@ -0,0 +1,220 @@
+
+		srvr := (*fakesrvrdata.FakeServerData)(fakeSrvrDataThs.Get())
+		srvr.System.ProcLoadAvg.CPU5m = newLoadAvg5M
+		fakeSrvrDataThs.Set(fakesrvrdata.ThsT(srvr))
```
Another `fakeSrvrDataThs.Set(srvr)` candidate

zrhoffman commented on traffic_monitor/tools/testcaches/fakesrvr/cmd.go:
```diff
@@ -0,0 +1,220 @@
+
+		srvr := (*fakesrvrdata.FakeServerData)(fakeSrvrDataThs.Get())
+		srvr.System.ProcLoadAvg.CPU10m = newLoadAvg10M
+		fakeSrvrDataThs.Set(fakesrvrdata.ThsT(srvr))
```
`fakeSrvrDataThs.Set(srvr)` can be here

zrhoffman commented on traffic_monitor/tools/testcaches/fakesrvr/server.go:
```diff
@@ -40,6 +43,20 @@ func reqIsApplicationSystem(r *http.Request) bool {
+		if maxDelayMS != 0 {
+			delayMS := minDelayMS
+			if minDelayMS != maxDelayMS {
+				delayMS += uint64(rand.Int63n(int64((maxDelayMS - minDelayMS))))
```
Redundant parenthesis here

zrhoffman commented on traffic_monitor/tests/integration/Dockerfile_run.sh:
```diff
@@ -0,0 +1,394 @@
+    "iloIpNetmask": null,
+    "iloPassword": null,
+    "iloUsername": null,
+    "interfaceMtu": null,
+    "interfaceName": "bond0",
+    "ip6Address": null,
+    "ip6Gateway": null,
+    "ipAddress": "trafficmonitor",
+    "ipGateway": "192.0.0.1",
+    "ipNetmask": "255.255.255.0",
```
After #4700, this part looks like

```json
      "interfaces": [
        {
          "ipAddresses": [
            {
              "address": "172.16.239.6",
              "gateway": "172.16.239.1",
              "serviceAddress": true
            },
            {
              "address": "fc01:9400:1000:8::6",
              "gateway": "fc01:9400:1000:8::1",
              "serviceAddress": true
            }
          ],
          "maxBandwidth": null,
          "monitor": true,
          "mtu": 1500,
          "name": "eth0"
        }
      ],
```

zrhoffman commented on traffic_monitor/tests/integration/Dockerfile_run.sh:
```diff
@@ -0,0 +1,394 @@
+}
+'
+
+	curl -Lvsk ${TESTTO_URI}/api/1.2/cdns/fake/configs/monitoring.json -X POST -d '
```
As with the rest of the API v2.0 routes, this has become `cdns/fake/configs/monitoring`, `json` suffix no longer supported

zrhoffman commented on traffic_monitor/tests/integration/README.md:
````diff
@@ -0,0 +1,12 @@
+From the `trafficcontrol/traffic_monitor` directory:
+
+```
+(cd tools/testto && go build)
+(cd tools/testcaches && go build)
+(cd tests/integration && go test -c -o traffic_monitor_integration_test)
````
When I run the docker-compose command to start the test,

```shell
docker-compose -p tmi --project-directory . -f tests/integration/docker-compose.yml run tmintegrationtest
```

I indefinitely get

```
Waiting for Traffic Ops to return a 200 OK
```

Looking at the `testto` container, it says

```logs
[user@computer traffic_monitor]$ 3 tmi --project-directory . -f tests/integration/docker-compose.yml logs -f testto
Attaching to tmi_testto_1
testto_1             | testto: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by testto)
tmi_testto_1 exited with code 1
```

So the instructions should have you build with `CGO_ENABLED=0` (and `GOOS=linux`).

zrhoffman commented on traffic_monitor/tests/integration/README.md:
```diff
@@ -0,0 +1,12 @@
+(cd tools/testto && go build)
+(cd tools/testcaches && go build)
+(cd tests/integration && go test -c -o traffic_monitor_integration_test)
+sudo docker-compose -p tmi --project-directory . -f tests/integration/docker-compose.yml run tmintegrationtest
```
After resolving the above linking error, rebuilding the images, and re-running, I get unmarshalling errors like

```json
{"error": "unmarshalling posted body: json: cannot unmarshal bool into Go struct field CRConfigDeliveryService.deliveryServices.anonymousBlockingEnabled of type string"}
```

It eventually says

```
Error communicating with Monitor 'http://trafficmonitor' - didn't return a 200 OK in 30s
```

<details><summary>Full <code>tmintegrationtest</code> logs are here (click to expand)</summary>

```
< HTTP/1.1 200 OK
* About to connect() to testto port 80 (#0)
*   Trying 172.23.0.3...
* Connected to testto (172.23.0.3) port 80 (#0)
> POST /api/1.2/cdns/fake/snapshot HTTP/1.1
> User-Agent: curl/7.29.0
> Host: testto
> Accept: */*
> Content-Length: 3756
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
< HTTP/1.1 400 Bad Request
< Date: Fri, 25 Sep 2020 21:32:19 GMT
< Content-Length: 169
< Content-Type: text/plain; charset=utf-8
* HTTP error before end of send, stop sending
<
* Closing connection 0
{"error": "unmarshalling posted body: json: cannot unmarshal bool into Go struct field CRConfigDeliveryService.deliveryServices.anonymousBlockingEnabled of type string"}* About to connect() to testto port 80 (#0)
*   Trying 172.23.0.3...
* Connected to testto (172.23.0.3) port 80 (#0)
> POST /api/1.2/cdns/fake/configs/monitoring.json HTTP/1.1
> User-Agent: curl/7.29.0
> Host: testto
> Accept: */*
> Content-Length: 2308
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
< HTTP/1.1 400 Bad Request
< Date: Fri, 25 Sep 2020 21:32:19 GMT
< Content-Length: 167
< Content-Type: text/plain; charset=utf-8
* HTTP error before end of send, stop sending
<
* Closing connection 0
{"error": "unmarshalling posted body: json: cannot unmarshal object into Go struct field TrafficServer.trafficServers.deliveryServices of type []tc.tsdeliveryService"}* About to connect() to testto port 80 (#0)
*   Trying 172.23.0.3...
* Connected to testto (172.23.0.3) port 80 (#0)
> POST /api/1.2/servers HTTP/1.1
> User-Agent: curl/7.29.0
> Host: testto
> Accept: */*
> Content-Length: 1135
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
< HTTP/1.1 204 No Content
< Date: Fri, 25 Sep 2020 21:32:19 GMT
<
* Connection #0 to host testto left intact


testto:
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3877    0  3877    0     0   657k      0 --:--:-- --:--:-- --:--:--  757k
{"response":{
  "config": {
    "api.cache-control.max_age": "30",
    "consistent.dns.routing": "true",
    "coveragezone.polling.interval": "30",


testcaches:
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
{ 0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
1  "ats": {
0    "plugin.remap_stats.num7.example.net.in_bytes": 2,
0    "plugin.remap_stats.num7.example.net.out_bytes": 2,
     "plugin.remap_stats.num7.example.net.status_2xx": 1,
34955    0 34955    0     0  4648k      0 --:--:-- --:--:-- --:--:-- 4876k
(23) Failed writing body


traffic_monitor:
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6522<!DOCTYPE html>0     0      0      0 --:--:-- --:--:-- --:--:--     0

 <!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
0  6522    0     0  1159k      0 --:--:-- --:--:-- --:--:-- 1273k
(23) Failed writing body
DEBUG traffic_monitor_integration starting

Error communicating with Monitor 'http://trafficmonitor' - didn't return a 200 OK in 30s
```
</details>

Looking at the logs of the `trafficmonitor` container, it doen't seem to get past

```go
ERROR: datareq.go:152: 2020-09-25T21:32:38.143874758Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
```

although there are other errors in there too, like

```go
trafficmonitor_1     | ERROR: monitorconfig.go:346: 2020-09-25T21:32:33.530687712Z: Failed to parse polling strings for cache server 'server0': no service addresses found
trafficmonitor_1     | ERROR: monitorconfig.go:346: 2020-09-25T21:32:33.530798803Z: Failed to parse polling strings for cache server 'server1': no service addresses found
```

<details><summary>Full <code>trafficmonitor</code> container logs are here (click to expand)</summary>

```go
Attaching to tmi_trafficmonitor_1
trafficmonitor_1     | Failed to get D-Bus connection: Operation not permitted
trafficmonitor_1     | /etc/init.d/traffic_monitor: line 41: /etc/sysconfig/network: No such file or directory
trafficmonitor_1     | Starting traffic_monitor: 
trafficmonitor_1     | ERROR: opsconfig.go:77: 2020-09-25T21:32:18.530467624Z: OpsConfigManager: getting CDN name from Traffic Ops, using config CDN 'nocdn': getting monitor CDN: no server 'trafficmonitor' found in Traffic Ops
trafficmonitor_1     | 
trafficmonitor_1     | ERROR: opsconfig.go:77: 2020-09-25T21:32:18.538113399Z: OpsConfigManager: Error getting Traffic Ops data: Error getting CRconfig from Traffic Ops: invalid CRConfig: CRConfig.Stats.CDN missing
trafficmonitor_1     | 
trafficmonitor_1     | ERROR: opsconfig.go:205: 2020-09-25T21:32:18.538157142Z: retrying in 100ms
trafficmonitor_1     | ERROR: opsconfig.go:77: 2020-09-25T21:32:18.638750564Z: OpsConfigManager: Error getting Traffic Ops data: Error getting CRconfig from Traffic Ops: invalid CRConfig: CRConfig.Stats.CDN missing
trafficmonitor_1     | 
trafficmonitor_1     | ERROR: opsconfig.go:205: 2020-09-25T21:32:18.638781915Z: retrying in 248.612746ms
trafficmonitor_1     | ERROR: opsconfig.go:77: 2020-09-25T21:32:18.888066827Z: OpsConfigManager: Error getting Traffic Ops data: Error getting CRconfig from Traffic Ops: invalid CRConfig: CRConfig.Stats.CDN missing
trafficmonitor_1     | 
trafficmonitor_1     | ERROR: opsconfig.go:205: 2020-09-25T21:32:18.888119514Z: retrying in 642.263625ms
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:20.102734056Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:21.1049107Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:22.107247084Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:23.109295834Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:24.111672602Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:25.113719758Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:26.115813436Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:27.118294393Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:28.120237324Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:29.122398443Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:30.123841491Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:31.125970238Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:32.128329372Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:33.130950123Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: monitorconfig.go:346: 2020-09-25T21:32:33.530687712Z: Failed to parse polling strings for cache server 'server0': no service addresses found
trafficmonitor_1     | ERROR: monitorconfig.go:346: 2020-09-25T21:32:33.530798803Z: Failed to parse polling strings for cache server 'server1': no service addresses found
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:34.135402114Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:35.136748588Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:36.139031306Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:37.141564084Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:38.143874758Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:39.146134866Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:40.14928317Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:41.151503741Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:42.154148808Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:43.156421926Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:44.158672414Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:45.159938856Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:46.161990084Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:47.163412929Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:48.164121248Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
trafficmonitor_1     | ERROR: datareq.go:152: 2020-09-25T21:32:49.165983499Z: Request Error: /api/version: service still starting, some caches unpolled: map[]
```
</details>

rob05c commented:
> @rob05c can this be closed in favor of #5817

Once it's merged, yeah. I'd prefer to keep it open until then so we don't lose track of it, just in case something happens to 5817 and it's closed or doesn't get merged or something

zrhoffman commented:
#5817 is based off of #3707 and is merged.
